### ISSUE TITLE
The implementation of wxRibbonToolBar::GetToolByPos(size_t pos) is broken

### ISSUE THREAD

**Issue by mcorino** [wxWidgets/wxWidgets#23457]
### Description
The implementation of `wxRibbonToolBar::GetToolByPos(size_t pos)` is broken for retrieving any tool item beyond the first group.

My intention was to iterate tool items in a `wxRibbonToolBar` by using `wxRibbonToolBar::GetToolCount()`  and `wxRibbonToolBar::GetToolByPos(size_t pos)` but after reading the code in `src/ribbon/toolbar.cpp` I had to conclude that that would not work although the code in `wxRibbonToolBar::GetToolPos(int tool_id)` made clear that it clearly was intended to work.

The code in `wxRibbonToolBar::GetToolPos(int tool_id)` correctly determines the position index for a tool item taking any preceding groups before the group containing the tool item  into account but the code in `wxRibbonToolBar::GetToolByPos(size_t pos)` fails to take those same group offsets into account so no position other than one from the first group will ever return a valid tool item.

To properly function  `wxRibbonToolBar::GetToolByPos(size_t pos)`  should be changed like this:

```C++
wxRibbonToolBarToolBase* wxRibbonToolBar::GetToolByPos(size_t pos)const
{
    size_t group_count = m_groups.GetCount();
    size_t g;
    for(g = 0; g < group_count; ++g)
    {
        wxRibbonToolBarToolGroup* group = m_groups.Item(g);
        size_t tool_count = group->tools.GetCount();
        if(pos<tool_count)
        {
            return group->tools[pos];
        }
        else if(pos==tool_count)
        {
            return nullptr;
        }
        pos -= tool_count+1;  // FIX CODE
    }
    return nullptr;
}
```

### Platform and version information

- wxWidgets version  you use: master
- wxWidgets port  you use: wxMSW
- OS  and its version: Windows 10

**Comment by vadz** [comment]
Sorry, but are you really sure this is the right fix? `pos` is unsigned, and the code above subtracts a value greater than it from it, why would this be the correct thing to do?

We need to add at least an interactive test for this to `samples/ribbon` (or, ideally, an automatic one to the test suite) and check that it really works.

Please don't hesitate to make PRs with the proposed changes.

**Comment by mcorino** [comment]
Where do you get the idea that `tool_count+1` is greater than `pos` at the moment of subtraction? There are two tests just before that make sure that when `pos <= tool_count` the code executes a `return`. At most `pos` can be equal to `tool_count+1` at the moment of subtraction in which the result is zero as it should be.

**Comment by mcorino** [comment]
Created a PR (#23458) with the proposed fix (which works locally) and a small test added to the ribbon sample.

**Comment by vadz** [comment]
Oops, sorry, I've somehow completely misread the code and "saw" `>` instead of `<`.

Anyhow, thanks for the PR, and especially for the test, will merge soon.

**Comment by mcorino** [comment]
When reviewing noticed that wxRibbonToolBar::DeleteToolByPos(size_t) suffers from the same problem so I added the same fix there (also in #23458).
